### PR TITLE
 Identify if only a subset of arguments is eager 

### PR DESCRIPTION
**Pull request by o-**
Try to analyze the order of argument forcing. Then use that
property to pass those arguments in an eager fashion.

### CODE REVIEW

File: rir/src/compiler/opt/force_dominance.cpp [hunk 1]
```diff
@@ -136,15 +133,63 @@ struct ForcedBy {
...
+        return ambiguousForceOrder;
+    }
+
+    bool eagerLikeFunction(ClosureVersion* fun) const {
+        if (!forceArgsBeforeEffect)
+            return false;
+        if (ambiguousForceOrder || argumentForceOrder.size() < fun->nargs())
+            return false;
+        for (size_t i = 0; i < fun->nargs(); ++i)
+            if (argumentForceOrder[i] != i)
```
Comment (mhyee) [inline rir/src/compiler/opt/force_dominance.cpp:182]: I'm trying to figure out what is going on here, and it seems like `eagerLikeFunction` requires _all_ arguments to be forced, in left-to-right order?

So if I had a function

```
f <- function(a1, a2, a3, a4) ...
```

the arguments need to be forced in the order `a1`, `a2`, `a3`, and `a4`?

And if they're force in some other order, e.g., `a2`, `a1`, `a3`, `a4`, that's no good?

File: rir/src/compiler/opt/force_dominance.cpp [hunk 2]
```diff
@@ -136,15 +133,63 @@ struct ForcedBy {
...
+                }
+            }
+        }
+
         return res;
     }
 
+    bool maybeForced(size_t i) const {
+        for (auto f : argumentForceOrder) {
+            if (f == i)
```
Comment (mhyee) [inline rir/src/compiler/opt/force_dominance.cpp:172]: This is to check if the given argument id `i` has already been forced?

Might be a bit heavy to bring in `<algorithm>` and `std::find`, but I think that would be slightly more clear. I definitely went "huh?" when I saw `if (f == i)`. Either way, a comment would help.

File: rir/src/compiler/opt/eager_calls.cpp [hunk 3]
```diff
@@ -28,20 +28,39 @@ void EagerCalls::apply(RirCompiler& cmp, ClosureVersion* closure,
...
+
+            auto isEager = [&](size_t i) {
+                // TODO: implement the case where we have a non-standard
+                // argument evaluation order
+                if (allEager)
+                    return true;
+                for (auto a : version->properties.argumentForceOrder) {
+                    if (a == i)
+                        return true;
+                    if (a > i)
```
Comment (mhyee) [inline rir/src/compiler/opt/eager_calls.cpp:46]: These index comparisons were also throwing me off, until I looked at `force_dominance.cpp`.

File: rir/src/compiler/opt/force_dominance.cpp (hunk 1 above)
Comment (o-) [reply rir/src/compiler/opt/force_dominance.cpp:182]: yes

**Pull request merged**